Sources/FoundationExtensions/Result+Extensions.swift

```diff
@@ -40,3 +40,21 @@ extension Result {
     }
 
 }
+
+extension Result where Success: OptionalType {
+
+    /// Converts a `Result<Success?, Error>` into `Result<Success, Error>?`
+    var asOptionalResult: Result<Success.Wrapped, Failure>? {
```

Comment:
I didn't write tests for this because the compiler is doing all the work checking the types. It's a simple type conversion with no added logic.

Reply:
I understand the point about the _current_ code being correct, but in practice, writing tests to check that the current code is correct is usually not that valuable in my experience. 
The real value in writing tests is future-proofing, so that if in the future another dev wants to modify or refactor this code, they can do so while resting assured that as long as tests pass, the code is still working correctly. 

Say that someone was debugging something, and in order to quickly test out what happens in the empty success case, they comment or remove the code in this implementation and replace it with a simple 

```swift
var asOptionalResult: Result<Success.Wrapped, Failure>? {
    return nil
}
```

Then they forget about it and never roll it back. Hopefully it would get caught in a PR, but if tests are passing, it might not be caught if it doesn't draw attention. 

If there are tests for this, it'll be easy to catch. Even if it doesn't make it to a full PR because other code is actually using the value and so the tests for that code fail, the tests will no longer be specific, and it might be hard to debug. 

So I'd always argue for making tests, even if the current code is guaranteed to be correct, in order to future-proof the code and ensure that other devs won't be able to accidentally break it.

Reply:
Always a great point that I keep forgetting. Added tests!